h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -1572,25 +1578,62 @@ private void fitIRLSM_multinomial(Solver s) {
       if (Solver.COORDINATE_DESCENT.equals(s)) {
         fitCOD_multinomial(s);
       } else {
-        double[] beta = _state.betaMultinomial();
+        double[] beta = _state.betaMultinomial(); // full with active/inactive columns
         do {
           beta = beta.clone();
           for (int c = 0; c < _nclass; ++c) {
             boolean onlyIcpt = _state.activeDataMultinomial(c).fullN() == 0;
             _state.setActiveClass(c);
-            LineSearchSolver ls = (_state.l1pen() == 0)
+            // _state.gslvrMultinomial(c) get beta, _state info, _state.betaMultinomial(c, beta) get coef per class
+            // _state.ginfoMultinomial(c) get gradient for one class
+            LineSearchSolver ls;
+            if (_parms._remove_collinear_columns && _state._iter < 1)
+              ls = (_state.l1pen() == 0)  // after first iteration over all classes, state._beta, ginfo.gradient are all smaller size
+                      ? new MoreThuente(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), _state.ginfoMultinomial(c))
+                      : new SimpleBacktrackingLS(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), _state.l1pen());
+            else if (_parms._remove_collinear_columns && _state._iter > 0)
+              ls = (_state.l1pen() == 0)  // after first iteration over all classes, state._beta, ginfo.gradient are all smaller size
+                      ? new MoreThuente(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), _state.ginfoMultinomialRCC(c))
+                      : new SimpleBacktrackingLS(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), _state.l1pen());
+            else
```

Comment:
To me, it is hard to see what is different between `iter == 0` and `iter > 0`. I'd find something like this more clear since there is a lot of duplication (but it's just a suggestion, maybe other people prefer it your way):
```suggestion
            if (_parms._remove_collinear_columns)
              ls = (_state.l1pen() == 0)  // after first iteration over all classes, state._beta, ginfo.gradient are all smaller size
                      ? new MoreThuente(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), (_state._iter < 1) ? _state.ginfoMultinomial(c) : _state.ginfoMultinomialRCC(c))
                      : new SimpleBacktrackingLS(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), _state.l1pen());
            else
```

Reply:
Tomas:  With rcc = true and iteration = 0, the dimension of beta and gradient per class has full size.  However, after the admm_solve, the beta and gradient per class will shrink.  So, for a class of 3 with full coefficients of 4, you will begin at 12 beta and 12 gradients.  Then,for class 0,  after first admm_solve, you will have beta of size 11 and gradient of size 11 because the first class has shrunk from 4 to 3.  Then, after you take care of class 1 and after admm_solve, the beta/gradient per class shrunk to 10.  For class 2, after admm_solve, beta/gradient shrunk to 9.  This will complete iteration 0.  Now, when you go to iteration 1, you will be dealing with beta/gradient per class of 9.  In addition, the _state.activeDataMultinomial() will have _adaptedFrame containing only of the active columns only while _state.activeDataMultinomial(c) will contain all the predictors in its _adaptedFrame.  I have to use different functions because at the beginning I can use  _state.activeDataMultinomial(c) since no columns have been made inactive.  However, after iteration 0 is done and at iteration 1, I will need to use  _state.activeDataMultinomial() since the only active columns are used.  

I think the best way to understand it is to go through the code line by line.  I used to have one function to deal with both cases but I made so many mistakes and decided to just write separate methods if I can to make things clean rather than using if/else statements in the methods.  Keeping track of all the beta/gradient dimension changes are exhausting....

Reply:
I think I got it but if I did you are missing the point of line 1600 in the suggested edits, namely the `(_state._iter < 1) ? _state.ginfoMultinomial(c) : _state.ginfoMultinomialRCC(c)`. The suggested edit was there just to make it more stand out what differs between the iteration 0 and iteration > 0.

It was just a suggestion that should IMO emphasize that this is the thing that is different because  in those 4 lines I couldn't easily spot the difference but then again you probably didn't in the suggested version so I either is probably fine...

Reply:
Tomas:  Yes, I missed your point but now I get it.  Good catch.  For some reason, git does not allow me to just commit your code.